arraycontext/container/traversal.py

```diff
@@ -256,43 +256,51 @@ def multimap_array_container(f: Callable[..., Any], *args: Any) -> Any:
 
 def rec_map_array_container(
         f: Callable[[Any], Any],
-        ary: ArrayOrContainerT) -> ArrayOrContainerT:
+        ary: ArrayOrContainerT,
+        leaf_class: Optional[type] = None) -> ArrayOrContainerT:
     r"""Applies *f* recursively to an :class:`ArrayContainer`.
 
     For a non-recursive version see :func:`map_array_container`.
 
     :param ary: a (potentially nested) structure of :class:`ArrayContainer`\ s,
         or an instance of a base array type.
     """
-    return _map_array_container_impl(f, ary, recursive=True)
+    return _map_array_container_impl(f, ary, leaf_cls=leaf_class, recursive=True)
 
 
 def mapped_over_array_containers(
-        f: Callable[[Any], Any]) -> Callable[[ArrayOrContainerT], ArrayOrContainerT]:
+        f: Callable[[Any], Any],
+        leaf_class: Optional[type] = None) -> Callable[
+            [ArrayOrContainerT], ArrayOrContainerT]:
     """Decorator around :func:`rec_map_array_container`."""
-    wrapper = partial(rec_map_array_container, f)
+    wrapper = partial(rec_map_array_container, f, leaf_class=leaf_class)
     update_wrapper(wrapper, f)
     return wrapper
```

Comment:
I don't think this is going to work as expected. Is it supposed to be used like
```python
@mapped_over_array_containers(leaf_class=DOFArray)
def my_function(ary):
	...
```
? In that case it should have another wrapper inside (there's an example in [here](https://www.pythontutorial.net/advanced-python/python-decorator-arguments/) for decorators with arguments).

Reply:
Whoops! Thanks, I'll fix it.

Reply:
Thoughts on b395fa827236cb9028fe85089b8dd7941b500630? Things ended up getting a little hairy with this; in order to preserve backwards compatibility and not have to add the `()` when not using `leaf_class=`, `[multi]mapped_over_array_containers` now behaves either as a decorator or decorator factory depending on how it's called. Not sure if this is OK?

Reply:
Yeah, looks good to me! At least I don't know of a cleaner way to do this sort of stuff.

Do these have a test to make sure they work in both cases?

Reply:
Doesn't look like it. I added some in 442a119f5a126cb31de5fabdfe2175da71b40b07.

Reply:
Doesn't look like it. I added some in 442a119f5a126cb31de5fabdfe2175da71b40b07.

Reply:
That's great! Thank you!